RFCs/Concepts_Redesign.md

```diff
@@ -0,0 +1,147 @@
+current concepts is that they need to be evaluated, evaluating a concept when
+trying to determine why it doesn't match your type can be really time consuming
+and frustrating. What this RFC proposes will make it significantly easier to
+understand what a concept specifies and to figure out why a type doesn't
+satisfy its spec.
+
+The downsides to this proposal are that what can be specified is diminished,
+that said I don't believe this is significant and I go through examples of this
+below.
+
+### Other proposals
+
+* https://github.com/nim-lang/RFCs/issues/13
+
+## Examples
+
+### Before
+
+Taken from: https://github.com/nim-lang/Nim/pull/12048#issuecomment-528093181
+
+```nim
+type
+  Iterable*[T] = concept c
+    for x in items(c): x is T
+    for x in mitems(c): x is var T
+
+  Indexable*[T] = concept c
+    type IndexType = type(c.low)
+    var i: IndexType
+
+    c[i] is T
+    c[i] is var T
+
+    var value: T
+    c[i] = value
+
+    c.len is IndexType
+```
+
+### After
+
+Taken from: https://github.com/nim-lang/Nim/pull/12048#issuecomment-529091854
+
+```nim
+type
+  Iterable*[T] = concept
+    iterator items(x: Iterable): T
+    iterator mitems(x: Iterable): var T
+  
+  Indexable*[T, IndexType] = concept
+    proc low(x: Indexable): IndexType
+    proc `[]`(x: Indexable, i: IndexType): T
+    proc `[]`(x: Indexable, i: IndexType): var T
+
+    proc `[]=`(x: Indexable, i: IndexType, val: T)
+    proc len(x: Indexable): IndexType
+```
+
+### Other cases
+
+#### Calls can refer to procs/templates/macros
+
+Inspired by https://github.com/nim-lang/Nim/pull/12048#issuecomment-529095046.
+
+From the example given by @bluenote10 above:
+
+> x.len could either refer to a field, a proc, a template, or a macro.
+> Even if we only allow proc, the actual signature could be different due
+> to optional arguments, varargs, or discarded return values.
+
+This can be solved by either softening what `proc` means in the concept or
+introducing a `callable` keyword:
+
+```nim
+type
+  Lengthable* = concept
+    callable len(x: Lengthable): int
+```
+
+#### Serializable concept
+
+Inspired by https://github.com/nim-lang/RFCs/issues/13#issuecomment-298886380.
+
+> Concept checking if a type is serializable by requiring that each field is serializable
+
+Before:
+
+```nim
+type Serializable = concept x
+   for f in fields(x):
+      f.serialize(Stream)
+```
+
+After:
+
+```nim
+type
+  Serializable[T] = concept
+    iterator fields(x: Serializable): T
+    proc serialize(f: T, s: Stream)
+```
+
+For the other examples suggested by @zah I would like to see more concrete
+use cases. I wouldn't be surprised if they can be implemented with the syntax
+I have in mind as well but I don't understand them enough to be able to
+provide equivalent examples.
+
+## Backward incompatibility
+
+This is a backwards incompatible change but concepts are an experimental feature
+which does not hold stability guarantees (even as of the upcoming v1.0).
```

Comment:
According to @Araq on IRC they do. (except for `do`)

Reply:
Link? Everything that's in experimental has no stability guarantees, we wrote this in the release notes.

Reply:
https://irclogs.nim-lang.org/06-09-2019.html#21:34:38

Reply:
Not that I necessarily agree that all experimental features should have stability guarantees..

Reply:
Er, to clarify, what I meant was these features won't be removed, but overhauled. 'do' is an exception, we can simply remove it, without loss of functionality.

Reply:
That's not true at all as far as our promise in the release notes goes. You shouldn't be making extra promises in such contexts, as far as the promise we made on nim-lang.org any experimental feature can be removed or replaced by another feature.